### PR TITLE
[KEYCLOAK-8348] - Containerize database tests

### PR DESCRIPTION
**Pull request by pedroigor**
The intent of this PR is run database tests using docker containers. Right now you can run tests using `mvn clean verify -Pmysql,jpa,auth-server-wildfly` or `mvn clean verify -Ppostgres,jpa,auth-server-wildfly`.

I'm finishing some last tests and how this works with CI. As well try to simplify the mvn comand to a single profile such as `-Pmysql|postgres`.

### CODE REVIEW

**Comment by vramik** [comment]
JPA matrix: https://keycloak-jenkins.com/job/test-keycloak-jpa/91/

**Comment by pedroigor** [comment]
@vramik, I have started a job with these changes plus other changes to pipeline itself in order to execute database tests using docker instead of db allocator.

The build was successful and I hope to push now the changes to pipeline earlier next week.

File: testsuite/integration-arquillian/pom.xml [hunk 1]
```diff
@@ -320,6 +320,157 @@
...
+        <profile>
+            <id>mysql</id>
+            <properties>
+                <keycloak.connectionsJpa.database>keycloak</keycloak.connectionsJpa.database>
+                <keycloak.connectionsJpa.user>keycloak</keycloak.connectionsJpa.user>
+                <keycloak.connectionsJpa.password>keycloak</keycloak.connectionsJpa.password>
+                <keycloak.connectionsJpa.url>jdbc:mysql://${auth.server.db.host}/${keycloak.connectionsJpa.database}</keycloak.connectionsJpa.url>
+                <jdbc.mvn.groupId>mysql</jdbc.mvn.groupId>
+                <jdbc.mvn.artifactId>mysql-connector-java</jdbc.mvn.artifactId>
+                <jdbc.mvn.version>5.1.29</jdbc.mvn.version>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: This should be `${mysql.version}`

File: testsuite/integration-arquillian/pom.xml [hunk 2]
```diff
@@ -320,6 +320,157 @@
...
+                <docker.database.image>mysql:5.7.25</docker.database.image>
+            </properties>
+            <build>
+                <pluginManagement>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: The plugins should be part of the generic flow, just plugin execution would be conditioned on existence of `docker.database.image`. The only issue is that the various containers need their specific env variables, e.g.  MYSQL_DATABASE, POSTGRES_DB - and all of those can be specified for `start-container` execution (rather renamed to `start-db-container` to allow other container types like LDAP to be handled similarly).

The profiles `db-postgres`, `db-mysql` would only define the `docker.database.image` and `jdbc.mvn.*` properties.

File: testsuite/integration-arquillian/pom.xml [hunk 3]
```diff
@@ -320,6 +320,157 @@
...
+                                </systemPropertyVariables>
+                            </configuration>
+                        </plugin>
+                        <plugin>
+                            <groupId>io.fabric8</groupId>
+                            <artifactId>docker-maven-plugin</artifactId>
+                            <version>0.28.0</version>
+                            <executions>
+                                <execution>
+                                    <id>start-container</id>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: `start-db-container`

File: testsuite/integration-arquillian/pom.xml [hunk 4]
```diff
@@ -320,6 +320,157 @@
...
+                                                        </tcp>
+                                                        <time>30000</time>
+                                                    </wait>
+                                                </run>
+                                            </image>
+                                        </images>
+                                    </configuration>
+                                </execution>
+                                <execution>
+                                    <id>stop-container</id>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: `stop-db-container`

File: testsuite/integration-arquillian/pom.xml [hunk 5]
```diff
@@ -320,6 +320,157 @@
...
+        <profile>
+            <id>postgres</id>
+            <properties>
+                <keycloak.connectionsJpa.database>keycloak</keycloak.connectionsJpa.database>
+                <keycloak.connectionsJpa.user>keycloak</keycloak.connectionsJpa.user>
+                <keycloak.connectionsJpa.password>keycloak</keycloak.connectionsJpa.password>
+                <keycloak.connectionsJpa.url>jdbc:postgresql://${auth.server.db.host}/${keycloak.connectionsJpa.database}</keycloak.connectionsJpa.url>
+                <jdbc.mvn.groupId>org.postgresql</jdbc.mvn.groupId>
+                <jdbc.mvn.artifactId>postgresql</jdbc.mvn.artifactId>
+                <jdbc.mvn.version>42.2.2</jdbc.mvn.version>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: `${postgresql.version}`

File: testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml [hunk 6]
```diff
@@ -127,6 +127,7 @@
                 ${kie.maven.settings}
                 -Dauth.server.truststore=${auth.server.truststore}
                 -Dauth.server.truststore.password=${auth.server.truststore.password}
+                -Dauth.server.db.host=${auth.server.db.host}
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml]: Why is this change necessary? Is it used in auth server at all? In such a case, there would be a change in XSLT updating server config.

If it is, there are many other places in this file where this has to be updated (every `auth-server-${auth.server}*`).

File: testsuite/integration-arquillian/tests/pom.xml [hunk 7]
```diff
@@ -1201,41 +1201,6 @@
             </build>
         </profile>
 
-
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/tests/pom.xml]: Do not remove this part. This is useful for non-docker databases.

**Review by hmlnarik — CHANGES_REQUESTED** [review]
Thanks for the PR. Please check the inline comments

File: testsuite/integration-arquillian/pom.xml [hunk 8]
```diff
@@ -320,6 +320,157 @@
...
+                                        <goal>stop</goal>
+                                    </goals>
+                                </execution>
+                            </executions>
+                        </plugin>
+                    </plugins>
+                </pluginManagement>
+            </build>
+        </profile>
+        <profile>
```
Comment (hmlnarik) [inline testsuite/integration-arquillian/pom.xml]: Profiles for `oracle` and `mssql` are needed too (it should only be matter of setting the properties)

**Comment by hmlnarik** [comment]
Please ensure that the PR runs on undertow auth server too.

File: testsuite/integration-arquillian/pom.xml (hunk 1 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/pom.xml]: "JDBC driver version" is the meaning of the `${mysql.version}`.

Comment (pedroigor) [reply testsuite/integration-arquillian/pom.xml]: I don't think the name is good. Besides, the `jdbc.mvn.version` is already in use by our testsuite to properly download and install the driver.

I tried to reuse what we already have in the testsuite as much as possible.

Comment (hmlnarik) [reply testsuite/integration-arquillian/pom.xml]: We have the same intention :) ${mysql.version} is already available (same for postgres), and I agree the name is unfortunate but it's already there: see e.g.
https://github.com/keycloak/keycloak/blob/master/pom.xml#L112
https://github.com/keycloak/keycloak/blob/master/testsuite/jetty/jetty93/pom.xml#L346

File: testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml (hunk 6 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml]: Yes it is. Just like the other system properties for truststore. As you noticed, we are passing the DB host.

Will update other auth-server-*.

Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml]: @pedroigor Where exactly? I could not find auth.server.db.host anywhere in the xsl sheets and in keycloak.connectionsJpa.url it is AFAICT expanded.

File: testsuite/integration-arquillian/pom.xml (hunk 1 above)
Comment (pedroigor) [reply testsuite/integration-arquillian/pom.xml]: Now I see what you mean, sorry. You don't want to change the property name but get the value from mysql.version.

Sometimes I get stuck with my own thoughts :)

File: testsuite/integration-arquillian/pom.xml (hunk 8 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/pom.xml]: If this PR would be ready in terms that the definition of a new database would be matter of creating a profile which would only consist of properties definition (`docker.database.image` alike) and adding env variables to docker run part, then yes, it can be done as a follow-up task.

File: testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml (hunk 6 above)
Comment (hmlnarik) [reply testsuite/integration-arquillian/tests/base/src/test/resources/arquillian.xml]: I still don't see where `auth.server.db.host` is used, even in the new commit (fc7c979). Can you point me to place where this is actually used within the auth server?

**Review by hmlnarik — APPROVED** [review]

**Comment by pedroigor** [comment]
@hmlnarik This happens sometimes, the job is not triggered by the bot. But it seems we have one now running. Lets wait ...
